bench/micro/Bench/Database/LSMTree/Internal/Merge.hs

```diff
@@ -0,0 +1,412 @@
+                !run <- merge hasFS conf p runs'
+                -- Make sure to immediately close resulting runs so we don't run
+                -- out of file handles. Ideally this would not be measured, but at
+                -- least it's pretty cheap.
+                Run.removeReference hasFS run
+        ]
+  where
+    withEnv =
+        Cr.envWithCleanup
+          (mergeEnv conf)
+          mergeEnvCleanup
+
+    -- We need to keep the input runs, but remove the freshly created one.
+    removeOutputRunFiles :: FS.HasFS IO FS.HandleIO -> IO ()
+    removeOutputRunFiles hasFS = do
+        traverse_ (FS.removeFile hasFS) (pathsForRunFiles outputRunPaths)
+        exists <- FS.doesFileExist hasFS (runChecksumsPath outputRunPaths)
+        when exists $
+          FS.removeFile hasFS (runChecksumsPath outputRunPaths)
+
+merge ::
+     FS.HasFS IO FS.HandleIO
+  -> Config
+  -> Run.RunFsPaths
+  -> InputRuns
+  -> IO (Run (FS.Handle (FS.HandleIO)))
+merge fs Config {..} targetPaths runs = do
+    let f = fromMaybe const mergeMappend
+    m <- fromMaybe (error "empty inputs, no merge created") <$>
+      Merge.new fs mergeLevel f targetPaths runs
+    go m
+  where
+    go m =
+        Merge.steps fs m stepSize >>= \case
+          Merge.MergeComplete run -> return run
+          Merge.MergeInProgress -> go m
+
+outputRunPaths :: Run.RunFsPaths
+outputRunPaths = RunFsPaths 0
+
+inputRunPaths :: [Run.RunFsPaths]
+inputRunPaths = RunFsPaths <$> [1..]
+
+type InputRuns = [Run (FS.Handle FS.HandleIO)]
+
+type Mappend = SerialisedValue -> SerialisedValue -> SerialisedValue
+
+onDeserialisedValues :: SerialiseValue v => (v -> v -> v) -> Mappend
+onDeserialisedValues f x y =
+    serialiseValue (f (deserialiseValue x) (deserialiseValue y))
+
+type SerialisedKOp = (SerialisedKey, SerialisedEntry)
+type SerialisedEntry = Entry SerialisedValue SerialisedBlob
+
+{-------------------------------------------------------------------------------
+  Environments
+-------------------------------------------------------------------------------}
+
+-- | Config options describing a benchmarking scenario
+data Config = Config {
+    -- | Name for the benchmark scenario described by this config.
+    name         :: !String
+    -- | Number of key\/operation pairs, one for each run.
+  , nentries     :: ![Int]
+    -- | Frequency of inserts within the key\/op pairs.
+  , finserts     :: !Int
+    -- | Frequency of inserts with blobs within the key\/op pairs.
+  , fblobinserts :: !Int
+    -- | Frequency of deletes within the key\/op pairs.
+  , fdeletes     :: !Int
+    -- | Frequency of mupserts within the key\/op pairs.
+  , fmupserts    :: !Int
+  , randomKey    :: Rnd SerialisedKey
+  , randomValue  :: Rnd SerialisedValue
+  , randomBlob   :: Rnd SerialisedBlob
+  , mergeLevel   :: !Merge.Level
+    -- | Needs to be defined when generating mupserts.
+  , mergeMappend :: !(Maybe Mappend)
```

Comment:
Looks like you use `const` by default when this is `Nothing`. So the comment doesn't seem entirely true

Reply:
I need to provide some function when creating the `Merge`, since there can always be mupserts in runs. But having `Nothing` here leads to `entryToUpdateNormal` being used in the input generation, which fails on mupserts.